File: shared/types/score.ts
// We will externalize file
// So business model (name, epoch, externalURL) remains shared,
// whilst the technical representation of the file
// is environment-specific.
// ==========
// React:
// ==========
// createScore(
//    payload: CreateScorePayload,
//    file: File,
// )
// ==========
// Vitest:
// ==========
// createScore(
//     payload: CreateScorePayload,
//     filePath: string,
// )

import { ComposerPublicResponse } from './composer';

export type CreateScorePayload = {
  composerId: number;
  scoreName: string;
  releaseDate?: string;
  categories?: string;
  tags?: string;
  informationText?: string;
  annotations?: string;
};

export type CreateScoreResponse = {
  message: string;
  id: number;
};

// ---------------------------
// Must correspond 100% to the dto !
export type ScorePublicResponse = {
  id: number;
  name: string;
  composerId: number;
  composer: ComposerPublicResponse;
  releaseDate: string;
  filepath: string;
  thumbnailPath: string;
  uploaderId: number;
  tags: string;
  categories: string;
  informationText: string;
  annotations: string;
  createdAt: string;
  updatedAt: string;
  isDemo: boolean;
};

// ---------------------------

export type GetScoresPageRequest = {
  page?: number;
  limit?: number;
  sort?: string;
  name?: string;
  search?: string;
  composer?: string;
  tag?: string;
  category?: string;
};

export type GetScoresPageResponse = {
  message: string;
  limit: number;
  page: number;
  sort?: string;
  total_rows: number;
  total_pages: number;
  scores: ScorePublicResponse[];
};

// ---------------------------

export type GetScoreRequest = {
  scoreId: number;
};

export type GetScoreResponse = {
  message: string;
  score: ScorePublicResponse;
};

// ---------------------------

export type UpdateScoreRequestPayload = {
  externalURL?: string;
  epoch?: string;
  isVerified?: boolean;
};

export type UpdateScoreResponse = {
  message: string;
  score: ScorePublicResponse;
};

// ---------------------------

export type Annotation = {
  id: string;
  page: number;
  type: string;
  geometry: {
    x: number;
    y: number;
    radius?: number;
    width?: number;
    height?: number;
  };
  style: {
    color: string;
    strokeWidth: number;
    opacity?: number;
  };
};

export type UpdateScoreAnnotationRequest = {
  scoreId: number;
  annotations: Annotation[];
};

export type UpdateScoreAnnotationResponse = {
  message: string;
};

// ---------------------------

File: backend/internal/dto/score.go
package dto

import "backend/internal/models"

type CreateScoreResponse struct {
	Message string `json:"message"`
	Id      uint32 `json:"id"`
}

type ScorePublicResponse struct {
	ID              uint32                 `json:"id"`
	ScoreName       string                 `json:"name"`
	ComposerID      uint32                 `json:"composerId"`
	Composer        ComposerPublicResponse `json:"composer"`
	ReleaseDate     string                 `json:"releaseDate"`
	FilePath        string                 `json:"filePath"`
	ThumbnailPath   string                 `json:"thumbnailPath"`
	UploaderID      uint32                 `json:"uploaderId"`
	Tags            string                 `json:"tags"`
	Categories      string                 `json:"categories"`
	InformationText string                 `json:"informationText"`
	Annotations     string                 `json:"annotations"`
	CreatedAt       string                 `json:"createdAt"`
	UpdatedAt       string                 `json:"updatedAt"`
	IsDemo          bool                   `json:"isDemo"`
}

type GetScoresPageResponse struct {
	Message    string                `json:"message"`
	Page       int                   `json:"page"`
	Limit      int                   `json:"limit"`
	TotalRows  int64                 `json:"total_rows"`
	TotalPages int                   `json:"total_pages"`
	Scores     []ScorePublicResponse `json:"scores"`
}

type GetScoreResponse struct {
	Message string              `json:"message"`
	Score   ScorePublicResponse `json:"score"`
}

// --------------------------------------------------------------------------
// Function
// --------------------------------------------------------------------------

func ToScorePublicResponse(score *models.Score) ScorePublicResponse {
	return ScorePublicResponse{
		ID:         score.ID,
		ScoreName:  score.ScoreName,
		ComposerID: score.ComposerID,

		Composer: ToComposerPublicResponse(&score.Composer),

		ReleaseDate:     score.ReleaseDate.Format("2006-01-02"),
		FilePath:        score.FilePath,
		ThumbnailPath:   score.ThumbnailPath,
		UploaderID:      score.UploaderID,
		Tags:            score.Tags,
		Categories:      score.Categories,
		InformationText: score.InformationText,
		Annotations:     score.Annotations,
		CreatedAt:       score.CreatedAt.Format("2006-01-02 15:04:05"),
		UpdatedAt:       score.UpdatedAt.Format("2006-01-02 15:04:05"),
		IsDemo:          score.IsDemo,
	}
}

func ToScoresPublicResponse(scores []*models.Score) []ScorePublicResponse {
	result := make([]ScorePublicResponse, len(scores))

	for i, score := range scores {
		result[i] = ToScorePublicResponse(score)
	}

	return result
}
